File: sdk/ts/test/synth-mcp-server.mjs
// Deterministic synthetic MCP server for deadline/cancellation/error tests
// (newline-delimited JSON-RPC 2.0 over stdio). Answers `initialize`
// immediately, then delays every other request by --delay-ms before replying.
// --exit-immediately exits at startup — the error-path probe.
//
// TEST fixture: deterministic timing so deadline/cancellation assertions
// cannot race a fast real server. Real edda round-trip/equivalence coverage
// stays in the live contract tests.

import args from "node:process";

const argv = args.argv.slice(2);
const flag = (name) => argv.includes(name);
const num = (name, fallback) => {
  const i = argv.indexOf(name);
  return i >= 0 ? Number(argv[i + 1]) : fallback;
};

if (flag("--exit-immediately")) process.exit(3);
const delayMs = num("--delay-ms", null);

import readline from "node:readline";
const rl = readline.createInterface({ input: process.stdin });
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

rl.on("line", (line) => {
  let msg;
  try {
    msg = JSON.parse(line);
  } catch {
    return;
  }
  if (msg.method === "initialize") {
    process.stdout.write(
      JSON.stringify({
        jsonrpc: "2.0",
        id: msg.id,
        result: {
          protocolVersion: "2025-03-26",
          capabilities: { tools: {} },
          serverInfo: { name: "synth", version: "0" },
        },
      }) + "\n",
    );
    return;
  }
  if (msg.id === null || msg.id === undefined) return; // notification
  void (async () => {
    if (delayMs != null) await sleep(delayMs);
    process.stdout.write(
      JSON.stringify({
        jsonrpc: "2.0",
        id: msg.id,
        result: { content: [{ type: "text", text: "synth" }] },
      }) + "\n",
    );
  })();
});

File: sdk/python/tests/synth_mcp_server.py
#!/usr/bin/env python3
"""Deterministic synthetic MCP server for deadline/cancellation/error tests.

Speaks newline-delimited JSON-RPC 2.0 over stdio (rmcp stdio framing):
answers `initialize` immediately, then delays every other request by
--delay seconds (default: never answers) before replying with a generic
result. --exit-immediately exits at startup instead — the error-path probe.

This is a TEST fixture: it deterministically controls timing so deadline and
cancellation assertions cannot race a fast real server. Real edda
round-trip/equivalence coverage stays in the live contract tests.
"""

import argparse
import json
import sys
import time


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--delay", type=float, default=None)
    ap.add_argument("--exit-immediately", action="store_true")
    args = ap.parse_args()
    if args.exit_immediately:
        sys.exit(3)
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
        except json.JSONDecodeError:
            continue
        if not isinstance(msg, dict):
            continue
        if msg.get("method") == "initialize":
            reply = {
                "jsonrpc": "2.0",
                "id": msg.get("id"),
                "result": {"protocolVersion": "2025-03-26", "capabilities": {"tools": {}}, "serverInfo": {"name": "synth", "version": "0"}},
            }
            sys.stdout.write(json.dumps(reply) + "\n")
            sys.stdout.flush()
            continue
        if "id" not in msg or msg.get("id") is None:
            continue  # notification
        if args.delay is not None:
            time.sleep(args.delay)
        sys.stdout.write(json.dumps({"jsonrpc": "2.0", "id": msg.get("id"), "result": {"content": [{"type": "text", "text": "synth"}]}}) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()
